### PR TITLE
Check for mixed new and old style imports

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lightning/issues/17485

cc @borda

### CODE REVIEW

File: src/lightning/pytorch/utilities/compile.py [hunk 1]
```diff
@@ -122,6 +123,7 @@ def _maybe_unwrap_optimized(model: object) -> "pl.LightningModule":
         return from_compiled(model)
     if isinstance(model, pl.LightningModule):
         return model
+    _check_mixed_imports(model)
```
Comment (awaelchli) [inline src/lightning/pytorch/utilities/compile.py:126]: why not include this check outside _maybe_unwrap_optimized()?  Seems unrelated to unwrapping.

Comment (carmocca) [reply src/lightning/pytorch/utilities/compile.py:126]: Just for the convenince of adding it to a single place. Currently, users who mix things up get the error just below

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]
